<a href="https://colab.research.google.com/github/darshandgithub/EDAprojects/blob/main/Heart_disease_prediction_by_classification.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
url='https://www.kaggle.com/datasets/yasserh/heart-disease-dataset'

In [ ]:
url

'https://www.kaggle.com/datasets/yasserh/heart-disease-dataset'

In [ ]:
! pip install opendatasets

In [ ]:
import opendatasets as od

In [ ]:
od.download(url)

Please provide your Kaggle credentials to download this dataset. Learn more: http://bit.ly/kaggle-creds
Your Kaggle username: Darshan D
Your Kaggle Key: ··········
Dataset URL: https://www.kaggle.com/datasets/yasserh/heart-disease-dataset


100%|██████████| 3.40k/3.40k [00:00<00:00, 2.14MB/s]

In [ ]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

In [ ]:
data=pd.read_csv('/content/heart-disease-dataset/heart.csv')

In [ ]:
data.head()

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope  \
0   63    1   3       145   233    1        0      150      0      2.3      0   
1   37    1   2       130   250    0        1      187      0      3.5      0   
2   41    0   1       130   204    0        0      172      0      1.4      2   
3   56    1   1       120   236    0        1      178      0      0.8      2   
4   57    0   0       120   354    0        1      163      1      0.6      2   

   ca  thal  target  
0   0     1       1  
1   0     2       1  
2   0     2       1  
3   0     2       1  
4   0     2       1

The dataset contains the following columns:

1. **age**: The age of the individual.
2. **sex**: Gender of the individual (1 = male, 0 = female).
3. **cp**: Chest pain type (0 = typical angina, 1 = atypical angina, 2 = non-anginal pain, 3 = asymptomatic).
4. **trestbps**: Resting blood pressure (in mm Hg).
5. **chol**: Serum cholesterol level (in mg/dl).
6. **fbs**: Fasting blood sugar > 120 mg/dl (1 = true, 0 = false).
7. **restecg**: Resting electrocardiographic results (0 = normal, 1 = ST-T wave abnormality, 2 = showing probable left ventricular hypertrophy).
8. **thalach**: Maximum heart rate achieved.
9. **exang**: Exercise-induced angina (1 = yes, 0 = no).
10. **oldpeak**: ST depression induced by exercise relative to rest.
11. **slope**: The slope of the peak exercise ST segment (0 = upsloping, 1 = flat, 2 = downsloping).
12. **ca**: Number of major vessels (0–3) colored by fluoroscopy.
13. **thal**: Thalassemia (1 = normal, 2 = fixed defect, 3 = reversible defect).
14. **target**: Diagnosis of heart disease (1 = presence, 0 = absence).

This dataset seems to be related to heart disease prediction.

In [ ]:
data.isnull().sum()

age         0
sex         0
cp          0
trestbps    0
chol        0
fbs         0
restecg     0
thalach     0
exang       0
oldpeak     0
slope       0
ca          0
thal        0
target      0
dtype: int64

In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 303 entries, 0 to 302
Data columns (total 14 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       303 non-null    int64  
 1   sex       303 non-null    int64  
 2   cp        303 non-null    int64  
 3   trestbps  303 non-null    int64  
 4   chol      303 non-null    int64  
 5   fbs       303 non-null    int64  
 6   restecg   303 non-null    int64  
 7   thalach   303 non-null    int64  
 8   exang     303 non-null    int64  
 9   oldpeak   303 non-null    float64
 10  slope     303 non-null    int64  
 11  ca        303 non-null    int64  
 12  thal      303 non-null    int64  
 13  target    303 non-null    int64  
dtypes: float64(1), int64(13)
memory usage: 33.3 KB


In [ ]:
from sklearn.model_selection import train_test_split

In [ ]:
train_df , test_df = train_test_split(data, test_size=0.2, random_state=42)

In [ ]:
train_df

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  \
132   42    1   1       120   295    0        1      162      0      0.0   
202   58    1   0       150   270    0        0      111      1      0.8   
196   46    1   2       150   231    0        1      147      0      3.6   
75    55    0   1       135   250    0        0      161      0      1.4   
176   60    1   0       117   230    1        1      160      1      1.4   
..   ...  ...  ..       ...   ...  ...      ...      ...    ...      ...   
188   50    1   2       140   233    0        1      163      0      0.6   
71    51    1   2        94   227    0        1      154      1      0.0   
106   69    1   3       160   234    1        0      131      0      0.1   
270   46    1   0       120   249    0        0      144      0      0.8   
102   63    0   1       140   195    0        1      179      0      0.0   

     slope  ca  thal  target  
132      2   0     2       1  
202      2   0     3       0  
196      1   0     2       0  
75       1   0     2       1  
176      2   2     3       0  
..     ...  ..   ...     ...  
188      1   1     3       0  
71       2   1     3       1  
106      1   1     2       1  
270      2   0     3       0  
102      2   2     2       1  

[242 rows x 14 columns]

In [ ]:
test_df

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  \
179   57    1   0       150   276    0        0      112      1      0.6   
228   59    1   3       170   288    0        0      159      0      0.2   
111   57    1   2       150   126    1        1      173      0      0.2   
246   56    0   0       134   409    0        0      150      1      1.9   
60    71    0   2       110   265    1        0      130      0      0.0   
..   ...  ...  ..       ...   ...  ...      ...      ...    ...      ...   
249   69    1   2       140   254    0        0      146      0      2.0   
104   50    1   2       129   196    0        1      163      0      0.0   
300   68    1   0       144   193    1        1      141      0      3.4   
193   60    1   0       145   282    0        0      142      1      2.8   
184   50    1   0       150   243    0        0      128      0      2.6   

     slope  ca  thal  target  
179      1   1     1       0  
228      1   0     3       0  
111      2   1     3       1  
246      1   2     3       0  
60       2   1     2       1  
..     ...  ..   ...     ...  
249      1   3     3       0  
104      2   0     2       1  
300      1   2     3       0  
193      1   2     3       0  
184      1   0     3       0  

[61 rows x 14 columns]

In [ ]:
data.shape

(303, 14)

In [ ]:
train_inputs = train_df.iloc[:,:13]
train_targets = train_df['target']
test_inputs = test_df.iloc[:,:13]
test_targets = test_df['target']

In [ ]:
train_inputs

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  \
132   42    1   1       120   295    0        1      162      0      0.0   
202   58    1   0       150   270    0        0      111      1      0.8   
196   46    1   2       150   231    0        1      147      0      3.6   
75    55    0   1       135   250    0        0      161      0      1.4   
176   60    1   0       117   230    1        1      160      1      1.4   
..   ...  ...  ..       ...   ...  ...      ...      ...    ...      ...   
188   50    1   2       140   233    0        1      163      0      0.6   
71    51    1   2        94   227    0        1      154      1      0.0   
106   69    1   3       160   234    1        0      131      0      0.1   
270   46    1   0       120   249    0        0      144      0      0.8   
102   63    0   1       140   195    0        1      179      0      0.0   

     slope  ca  thal  
132      2   0     2  
202      2   0     3  
196      1   0     2  
75       1   0     2  
176      2   2     3  
..     ...  ..   ...  
188      1   1     3  
71       2   1     3  
106      1   1     2  
270      2   0     3  
102      2   2     2  

[242 rows x 13 columns]

In [ ]:
test_inputs

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  \
179   57    1   0       150   276    0        0      112      1      0.6   
228   59    1   3       170   288    0        0      159      0      0.2   
111   57    1   2       150   126    1        1      173      0      0.2   
246   56    0   0       134   409    0        0      150      1      1.9   
60    71    0   2       110   265    1        0      130      0      0.0   
..   ...  ...  ..       ...   ...  ...      ...      ...    ...      ...   
249   69    1   2       140   254    0        0      146      0      2.0   
104   50    1   2       129   196    0        1      163      0      0.0   
300   68    1   0       144   193    1        1      141      0      3.4   
193   60    1   0       145   282    0        0      142      1      2.8   
184   50    1   0       150   243    0        0      128      0      2.6   

     slope  ca  thal  
179      1   1     1  
228      1   0     3  
111      2   1     3  
246      1   2     3  
60       2   1     2  
..     ...  ..   ...  
249      1   3     3  
104      2   0     2  
300      1   2     3  
193      1   2     3  
184      1   0     3  

[61 rows x 13 columns]

In [ ]:
from sklearn.linear_model import LogisticRegression

In [ ]:
lg_model = LogisticRegression(solver='saga')

In [ ]:
lg_model.fit(train_inputs, train_targets)

/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_sag.py:349: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn(


LogisticRegression(solver='saga')

In [ ]:
predictions_lg = lg_model.predict(test_inputs)

In [ ]:
test_targets

179    0
228    0
111    1
246    0
60     1
      ..
249    0
104    1
300    0
193    0
184    0
Name: target, Length: 61, dtype: int64

In [ ]:
predictions_lg

array([0, 0, 1, 0, 0, 1, 1, 0, 0, 1, 1, 0, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0,
       0, 0, 1, 1, 0, 1, 0, 1, 0, 0, 0, 1, 0, 0, 1, 1, 1, 1, 1, 1, 0, 1,
       1, 1, 1, 1, 0, 0, 1, 1, 1, 1, 1, 0, 0, 1, 0, 0, 0])

In [ ]:
from sklearn.metrics import accuracy_score

In [ ]:
accuracy_score(test_targets, predictions_lg)

0.7704918032786885

In [ ]:
from sklearn.metrics import confusion_matrix

In [ ]:
confusion_matrix(test_targets, predictions_lg)

array([[22,  7],
       [ 7, 25]])

##Decision tree

In [ ]:
from sklearn.tree import DecisionTreeClassifier

In [ ]:
dt_model = DecisionTreeClassifier()

In [ ]:
dt_model.fit(train_inputs, train_targets)

DecisionTreeClassifier()

In [ ]:
predictions_dt = dt_model.predict(test_inputs)

In [ ]:
accuracy_score(test_targets, predictions_dt)

0.8032786885245902

#Random Forest

In [ ]:
from sklearn.ensemble import RandomForestClassifier

In [ ]:
rf_model = RandomForestClassifier()

In [ ]:
rf_model.fit(train_inputs, train_targets)

RandomForestClassifier()

In [ ]:
prdictions_rf = rf_model.predict(test_inputs)

In [ ]:
accuracy_score(test_targets, prdictions_rf)

0.819672131147541

In [ ]:
data.head()

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope  \
0   63    1   3       145   233    1        0      150      0      2.3      0   
1   37    1   2       130   250    0        1      187      0      3.5      0   
2   41    0   1       130   204    0        0      172      0      1.4      2   
3   56    1   1       120   236    0        1      178      0      0.8      2   
4   57    0   0       120   354    0        1      163      1      0.6      2   

   ca  thal  target  
0   0     1       1  
1   0     2       1  
2   0     2       1  
3   0     2       1  
4   0     2       1

In [ ]:
rf_model.predict([[63	,1	,3	,145,	233	,1,	0	,150,	0,	2.3,	0,	0,	1]])

/usr/local/lib/python3.10/dist-packages/sklearn/base.py:493: UserWarning: X does not have valid feature names, but RandomForestClassifier was fitted with feature names
  warnings.warn(


array([1])

In [ ]:
from sklearn.metrics import classification_report

In [ ]:
print(classification_report(test_targets, prdictions_rf))

              precision    recall  f1-score   support

           0       0.80      0.83      0.81        29
           1       0.84      0.81      0.83        32

    accuracy                           0.82        61
   macro avg       0.82      0.82      0.82        61
weighted avg       0.82      0.82      0.82        61



In [ ]:
data

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  \
0     63    1   3       145   233    1        0      150      0      2.3   
1     37    1   2       130   250    0        1      187      0      3.5   
2     41    0   1       130   204    0        0      172      0      1.4   
3     56    1   1       120   236    0        1      178      0      0.8   
4     57    0   0       120   354    0        1      163      1      0.6   
..   ...  ...  ..       ...   ...  ...      ...      ...    ...      ...   
298   57    0   0       140   241    0        1      123      1      0.2   
299   45    1   3       110   264    0        1      132      0      1.2   
300   68    1   0       144   193    1        1      141      0      3.4   
301   57    1   0       130   131    0        1      115      1      1.2   
302   57    0   1       130   236    0        0      174      0      0.0   

     slope  ca  thal  target  
0        0   0     1       1  
1        0   0     2       1  
2        2   0     2       1  
3        2   0     2       1  
4        2   0     2       1  
..     ...  ..   ...     ...  
298      1   0     3       0  
299      1   0     3       0  
300      1   2     3       0  
301      1   1     3       0  
302      1   1     2       0  

[303 rows x 14 columns]

In [ ]:
data.target.unique()

array([1, 0])

In [ ]:
!pip install keras

In [ ]:
from sklearn.preprocessing import OneHotEncoder
from tensorflow.keras.utils import to_categorical
from sklearn.preprocessing import LabelEncoder

In [ ]:
x = data.iloc[:,:-1]

In [ ]:
x

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  \
0     63    1   3       145   233    1        0      150      0      2.3   
1     37    1   2       130   250    0        1      187      0      3.5   
2     41    0   1       130   204    0        0      172      0      1.4   
3     56    1   1       120   236    0        1      178      0      0.8   
4     57    0   0       120   354    0        1      163      1      0.6   
..   ...  ...  ..       ...   ...  ...      ...      ...    ...      ...   
298   57    0   0       140   241    0        1      123      1      0.2   
299   45    1   3       110   264    0        1      132      0      1.2   
300   68    1   0       144   193    1        1      141      0      3.4   
301   57    1   0       130   131    0        1      115      1      1.2   
302   57    0   1       130   236    0        0      174      0      0.0   

     slope  ca  thal  
0        0   0     1  
1        0   0     2  
2        2   0     2  
3        2   0     2  
4        2   0     2  
..     ...  ..   ...  
298      1   0     3  
299      1   0     3  
300      1   2     3  
301      1   1     3  
302      1   1     2  

[303 rows x 13 columns]

In [ ]:
y = data['target']

In [ ]:
y

0      1
1      1
2      1
3      1
4      1
      ..
298    0
299    0
300    0
301    0
302    0
Name: target, Length: 303, dtype: int64

In [ ]:
from sklearn.model_selection import train_test_split

In [ ]:
x_train, x_test, y_train, y_test = train_test_split(x,y, test_size=0.2, random_state=42)

In [ ]:
y_train_onehot = to_categorical(y_train, 2)
y_test_onehot = to_categorical(y_train, 2)

In [ ]:
from keras.models import Sequential
from keras.layers import Dense
from keras.optimizers import Adam

In [ ]:
model = Sequential()

In [ ]:
model.add(Dense(units=30, activation='relu', input_shape=(13,)))

/usr/local/lib/python3.10/dist-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


In [ ]:
model.add(Dense(units=15, activation='relu'))

In [ ]:
model.add(Dense(units=2, activation='softmax'))

In [ ]:
model.compile(optimizer='adam',loss='categorical_crossentropy', metrics=['accuracy'])

In [ ]:
model.fit(x_train,y_train_onehot, epochs=60,validation_split=0.1, batch_size=16)

Epoch 1/60
14/14 ━━━━━━━━━━━━━━━━━━━━ 0s 13ms/step - accuracy: 0.4716 - loss: 0.6943 - val_accuracy: 0.6000 - val_loss: 0.6238
Epoch 2/60
14/14 ━━━━━━━━━━━━━━━━━━━━ 0s 9ms/step - accuracy: 0.5589 - loss: 0.6698 - val_accuracy: 0.6400 - val_loss: 0.6683
Epoch 3/60
14/14 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step - accuracy: 0.6241 - loss: 0.6630 - val_accuracy: 0.6400 - val_loss: 0.6657
Epoch 4/60
14/14 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step - accuracy: 0.5247 - loss: 0.6669 - val_accuracy: 0.6000 - val_loss: 0.6446
Epoch 5/60
14/14 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.5166 - loss: 0.6677 - val_accuracy: 0.6400 - val_loss: 0.6228
Epoch 6/60
14/14 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.5599 - loss: 0.6583 - val_accuracy: 0.6800 - val_loss: 0.6583
Epoch 7/60
14/14 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step - accuracy: 0.5959 - loss: 0.6223 - val_accuracy: 0.6800 - val_loss: 0.6049
Epoch 8/60
14/14 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step - accuracy: 0.5910 - loss: 0.6606 - val_accuracy: 0.5200 - val_loss

In [ ]:
model.predict(x_test)

2/2 ━━━━━━━━━━━━━━━━━━━━ 1s 491ms/step


array([[0.5142715 , 0.48572856],
       [0.5142715 , 0.48572856],
       [0.23058219, 0.7694179 ],
       [0.5142715 , 0.48572856],
       [0.5142715 , 0.48572856],
       [0.23058219, 0.7694179 ],
       [0.5142715 , 0.48572856],
       [0.5142715 , 0.48572856],
       [0.5142715 , 0.48572856],
       [0.5142715 , 0.48572856],
       [0.51425123, 0.4857488 ],
       [0.5142715 , 0.48572856],
       [0.5142715 , 0.48572856],
       [0.5142713 , 0.48572862],
       [0.5071208 , 0.49287927],
       [0.49804983, 0.50195014],
       [0.23058222, 0.76941776],
       [0.5142715 , 0.48572856],
       [0.5142715 , 0.48572856],
       [0.5142712 , 0.48572874],
       [0.5142715 , 0.48572856],
       [0.5142715 , 0.48572856],
       [0.5142715 , 0.48572856],
       [0.5142715 , 0.48572856],
       [0.5142715 , 0.48572856],
       [0.23059513, 0.76940495],
       [0.5142715 , 0.48572856],
       [0.5142715 , 0.48572856],
       [0.5142715 , 0.48572856],
       [0.5142715 , 0.48572856],
       [0.